test/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/SelfEmployedCYAControllerSpec.scala

```diff
@@ -0,0 +1,183 @@
+    "submit" -> TestSelfEmployedCYAController.submit(id)
+  )
+
+  override def beforeEach(): Unit = {
+    disable(SaveAndRetrieve)
+    super.beforeEach()
+  }
+
+  object TestSelfEmployedCYAController extends SelfEmployedCYAController(
+    selfEmployedCYA,
+    mockAuthService,
+    mockIncomeTaxSubscriptionConnector,
+    mockMultipleSelfEmploymentsService,
+    mockMessagesControllerComponents
+  )
+
+  val selfEmployment: SelfEmploymentData = SelfEmploymentData(
+    id = id,
+    businessStartDate = Some(BusinessStartDate(DateModel("1", "1", "1980"))),
+    businessName = Some(BusinessNameModel("testBusinessName")),
+    businessTradeName = Some(BusinessTradeNameModel("testBusinessTrade")),
+    businessAddress = Some(BusinessAddressModel(auditRef = "testAuditRef", address = Address(lines = Seq("line 1"), postcode = "ZZ1 1ZZ"))),
+    addressRedirect = Some("/test-redirect")
+  )
+
+  val incompleteSelfEmployment: SelfEmploymentData = SelfEmploymentData(
+    id = id,
+    businessStartDate = Some(BusinessStartDate(DateModel("1", "1", "1980"))),
+    businessName = Some(BusinessNameModel("testBusinessName")),
+    businessTradeName = None,
+    businessAddress = Some(BusinessAddressModel(auditRef = "testAuditRef", address = Address(lines = Seq("line 1"), postcode = "ZZ1 1ZZ"))),
+    addressRedirect = Some("/test-redirect")
+  )
+
+  "show" when {
+    "the save and retrieve feature switch is disabled" should {
+      "throw a not found exception" in {
+        intercept[NotFoundException](await(TestSelfEmployedCYAController.show(id, isEditMode = false)(fakeRequest)))
+          .message mustBe "[SelfEmployedCYAController][show] - The save and retrieve feature switch is disabled"
+      }
+    }
+    "the save and retrieve feature switch is enabled" should {
+      "throw an internal server exception" when {
+        "there was a problem retrieving the users accounting method" in {
+          enable(SaveAndRetrieve)
+
+          mockGetSelfEmployments(businessAccountingMethodKey)(Left(UnexpectedStatusFailure(INTERNAL_SERVER_ERROR)))
+
+          intercept[InternalServerException](await(TestSelfEmployedCYAController.show(id, isEditMode = false)(fakeRequest)))
+            .message mustBe "[SelfEmployedCYAController][withSelfEmploymentCYAModel] - Failure retrieving accounting method"
+        }
+        "there was a problem retrieving the users self employment data" in {
+          enable(SaveAndRetrieve)
+          mockGetSelfEmployments[AccountingMethodModel](businessAccountingMethodKey)(Right(Some(AccountingMethodModel(Cash))))
+          mockFetchBusiness(id)(Left(UnexpectedStatusFailure(INTERNAL_SERVER_ERROR)))
+
+          intercept[InternalServerException](await(TestSelfEmployedCYAController.show(id, isEditMode = false)(fakeRequest)))
+            .message mustBe "[SelfEmployedCYAController][withSelfEmploymentCYAModel] - Failure retrieving self employment data"
+        }
+      }
+      "return OK" when {
```

Comment:
This test is failing with a `java.lang.NullPointerException`, you can fix it with:
```
      "return OK" when {
        "all data is retrieved successfully" in withController { controller =>
          enable(SaveAndRetrieve)
          mockGetSelfEmployments[AccountingMethodModel](businessAccountingMethodKey)(Right(Some(AccountingMethodModel(Cash))))
          mockFetchBusiness(id)(Right(Some(selfEmployment)))

          val result: Future[Result] = controller.show(id)(fakeRequest)

          status(result) mustBe OK
          contentType(result) mustBe Some(HTML)
        }
      }
```
and add the following private function at the end of the `SelfEmployedCYAControllerSpec` class:
```
  private def withController(testCode: SelfEmployedCYAController => Any) = {
    val selfEmployedCYA: SelfEmployedCYA = mock[SelfEmployedCYA]

    when(selfEmployedCYA(any(), any(), any())(any(), any()))
      .thenReturn(HtmlFormat.empty)

    val controller = new SelfEmployedCYAController(
      selfEmployedCYA,
      mockAuthService,
      mockIncomeTaxSubscriptionConnector,
      mockMultipleSelfEmploymentsService,
      mockMessagesControllerComponents
    )

    testCode(controller)
  }
```